File: sdks/typescript/src/session.ts
/**
 * Session Management for AvocadoDB TypeScript SDK
 *
 * Provides conversation session management with:
 * - Multi-turn conversation tracking
 * - Context compilation in session context
 * - Conversation history retrieval
 * - Session replay for debugging
 *
 * @example
 * ```typescript
 * const db = new AvocadoDB('http://localhost:8765');
 *
 * // Create a session
 * const session = await db.createSession({
 *   userId: 'alice',
 *   title: 'Project Q&A'
 * });
 *
 * // Multi-turn conversation
 * const result = await session.compile('What is AvocadoDB?');
 * await session.addMessage('assistant', 'AvocadoDB is...');
 *
 * // Get conversation history
 * const history = await session.getHistory();
 *
 * // Replay for debugging
 * const replay = await session.replay();
 * ```
 */

import { WorkingSet } from './client';

/**
 * Message role in a conversation
 */
export type MessageRole = 'user' | 'assistant' | 'system' | 'tool';

/**
 * A message in a conversation session
 */
export interface Message {
  /** Message ID */
  id: string;

  /** Session ID */
  sessionId: string;

  /** Message role */
  role: MessageRole;

  /** Message content */
  content: string;

  /** Sequence number (0-indexed) */
  sequenceNumber: number;

  /** Creation timestamp */
  createdAt: string;

  /** Optional metadata */
  metadata?: Record<string, any>;
}

/**
 * Session metadata
 */
export interface SessionInfo {
  /** Session ID */
  id: string;

  /** Optional user identifier */
  userId?: string;

  /** Optional session title */
  title?: string;

  /** Creation timestamp */
  createdAt: string;

  /** Last update timestamp */
  updatedAt: string;

  /** Last message timestamp */
  lastMessageAt?: string;

  /** Optional metadata */
  metadata?: Record<string, any>;
}

/**
 * A conversation turn (user query + optional assistant response)
 */
export interface SessionTurn {
  /** User message */
  userMessage: Message;

  /** Compiled context (if any) */
  workingSet?: WorkingSet;

  /** Assistant response (if any) */
  assistantMessage?: Message;
}

/**
 * Session replay data for debugging
 */
export interface SessionReplay {
  /** Session metadata */
  session: SessionInfo;

  /** Conversation turns */
  turns: SessionTurn[];
}

/**
 * Options for creating a session
 */
export interface CreateSessionOptions {
  /** Optional user identifier */
  userId?: string;

  /** Optional session title */
  title?: string;

  /** Optional metadata */
  metadata?: Record<string, any>;
}

/**
 * Options for compiling in session context
 */
export interface SessionCompileOptions {
  /** Token budget (default: 8000) */
  budget?: number;

  /** Optional metadata */
  metadata?: Record<string, any>;
}

/**
 * Result of session compilation
 */
export interface SessionCompileResult {
  /** User message that was added */
  message: Message;

  /** Compiled context */
  workingSet: WorkingSet;
}

/**
 * Session class for managing conversations
 */
export class Session {
  private baseUrl: string;
  private project: string;

  /** Session ID */
  readonly id: string;

  /** User ID (if set) */
  userId?: string;

  /** Session title (if set) */
  title?: string;

  /** Creation timestamp */
  readonly createdAt: string;

  /** Last update timestamp */
  updatedAt: string;

  constructor(baseUrl: string, project: string, data: SessionInfo) {
    this.baseUrl = baseUrl;
    this.project = project;
    this.id = data.id;
    this.userId = data.userId;
    this.title = data.title;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
  }

  /**
   * Add a message to the session
   *
   * @param role - Message role
   * @param content - Message content
   * @param metadata - Optional metadata
   * @returns The created message
   */
  async addMessage(
    role: MessageRole,
    content: string,
    metadata?: Record<string, any>
  ): Promise<Message> {
    const response = await fetch(
      `${this.baseUrl}/sessions/${this.id}/messages`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          role,
          content,
          metadata,
          project: this.project,
        }),
      }
    );

    if (!response.ok) {
      throw new Error(`Failed to add message: ${response.statusText}`);
    }

    const data = (await response.json()) as { message: any };
    return this.parseMessage(data.message);
  }

  /**
   * Compile context for a query in session context
   *
   * This automatically adds a user message and compiles context.
   *
   * @param query - User query
   * @param options - Compilation options
   * @returns Compilation result
   */
  async compile(
    query: string,
    options: SessionCompileOptions = {}
  ): Promise<SessionCompileResult> {
    const response = await fetch(
      `${this.baseUrl}/sessions/${this.id}/compile`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          query,
          token_budget: options.budget || 8000,
          metadata: options.metadata,
          project: this.project,
        }),
      }
    );

    if (!response.ok) {
      throw new Error(`Failed to compile: ${response.statusText}`);
    }

    const data = (await response.json()) as { message: any; working_set: any };

    return {
      message: this.parseMessage(data.message),
      workingSet: new WorkingSet(data.working_set),
    };
  }

  /**
   * Get formatted conversation history
   *
   * @param maxTokens - Optional token limit (keeps recent messages)
   * @returns Formatted conversation history
   */
  async getHistory(maxTokens?: number): Promise<string> {
    const url = new URL(`${this.baseUrl}/sessions/${this.id}/history`);
    url.searchParams.append('project', this.project);

    if (maxTokens) {
      url.searchParams.append('max_tokens', maxTokens.toString());
    }

    const response = await fetch(url.toString());

    if (!response.ok) {
      throw new Error(`Failed to get history: ${response.statusText}`);
    }

    const data = (await response.json()) as { history: string };
    return data.history;
  }

  /**
   * Replay session for debugging
   *
   * Returns structured data about each conversation turn,
   * including compiled context and citations.
   *
   * @returns Session replay data
   */
  async replay(): Promise<SessionReplay> {
    const url = new URL(`${this.baseUrl}/sessions/${this.id}/replay`);
    url.searchParams.append('project', this.project);

    const response = await fetch(url.toString());

    if (!response.ok) {
      throw new Error(`Failed to replay session: ${response.statusText}`);
    }

    const data = (await response.json()) as { session: any; turns: any[] };

    return {
      session: this.parseSessionInfo(data.session),
      turns: data.turns.map((turn: any) => ({
        userMessage: this.parseMessage(turn.user_message),
        workingSet: turn.working_set
          ? new WorkingSet(turn.working_set)
          : undefined,
        assistantMessage: turn.assistant_message
          ? this.parseMessage(turn.assistant_message)
          : undefined,
      })),
    };
  }

  /**
   * Delete this session
   */
  async delete(): Promise<void> {
    const url = new URL(`${this.baseUrl}/sessions/${this.id}`);
    url.searchParams.append('project', this.project);

    const response = await fetch(url.toString(), {
      method: 'DELETE',
    });

    if (!response.ok) {
      throw new Error(`Failed to delete session: ${response.statusText}`);
    }
  }

  /**
   * Refresh session data from server
   */
  async refresh(): Promise<void> {
    const url = new URL(`${this.baseUrl}/sessions/${this.id}`);
    url.searchParams.append('project', this.project);

    const response = await fetch(url.toString());

    if (!response.ok) {
      throw new Error(`Failed to refresh session: ${response.statusText}`);
    }

    const data = (await response.json()) as { session: any };
    const sessionData = this.parseSessionInfo(data.session);

    this.userId = sessionData.userId;
    this.title = sessionData.title;
    this.updatedAt = sessionData.updatedAt;
  }

  /**
   * Get a string representation of the session
   */
  toString(): string {
    return `Session(id=${this.id}, user=${this.userId || 'N/A'}, title=${this.title || 'N/A'})`;
  }

  private parseMessage(data: any): Message {
    return {
      id: data.id,
      sessionId: data.session_id,
      role: data.role,
      content: data.content,
      sequenceNumber: data.sequence_number,
      createdAt: data.created_at,
      metadata: data.metadata,
    };
  }

  private parseSessionInfo(data: any): SessionInfo {
    return {
      id: data.id,
      userId: data.user_id,
      title: data.title,
      createdAt: data.created_at,
      updatedAt: data.updated_at,
      lastMessageAt: data.last_message_at,
      metadata: data.metadata,
    };
  }
}

/**
 * Session management methods for AvocadoDB client
 *
 * These methods should be added to the main AvocadoDB class.
 */
export class SessionManager {
  constructor(private baseUrl: string, private project: string) {}

  /**
   * Create a new session
   *
   * @param options - Session options
   * @returns The created session
   */
  async createSession(options: CreateSessionOptions = {}): Promise<Session> {
    const response = await fetch(`${this.baseUrl}/sessions`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        user_id: options.userId,
        title: options.title,
        metadata: options.metadata,
        project: this.project,
      }),
    });

    if (!response.ok) {
      throw new Error(`Failed to create session: ${response.statusText}`);
    }

    const data = (await response.json()) as { session: any };
    return new Session(this.baseUrl, this.project, data.session);
  }

  /**
   * List sessions
   *
   * @param userId - Optional user ID filter
   * @param limit - Maximum number of sessions to return
   * @returns Array of session info
   */
  async listSessions(
    userId?: string,
    limit: number = 50
  ): Promise<SessionInfo[]> {
    const url = new URL(`${this.baseUrl}/sessions`);
    url.searchParams.append('project', this.project);

    if (userId) {
      url.searchParams.append('user_id', userId);
    }

    url.searchParams.append('limit', limit.toString());

    const response = await fetch(url.toString());

    if (!response.ok) {
      throw new Error(`Failed to list sessions: ${response.statusText}`);
    }

    const data = (await response.json()) as { sessions: any[] };
    return data.sessions.map((s: any) => ({
      id: s.id,
      userId: s.user_id,
      title: s.title,
      createdAt: s.created_at,
      updatedAt: s.updated_at,
      lastMessageAt: s.last_message_at,
      metadata: s.metadata,
    }));
  }

  /**
   * Get a session by ID
   *
   * @param sessionId - Session ID
   * @returns The session
   */
  async getSession(sessionId: string): Promise<Session> {
    const url = new URL(`${this.baseUrl}/sessions/${sessionId}`);
    url.searchParams.append('project', this.project);

    const response = await fetch(url.toString());

    if (!response.ok) {
      throw new Error(`Failed to get session: ${response.statusText}`);
    }

    const data = (await response.json()) as { session: any };
    return new Session(this.baseUrl, this.project, data.session);
  }
}

File: sdks/python/avocado/session.py
"""
Session management for AvocadoDB.

This module provides a high-level Session class for managing conversation
sessions, including message management, context compilation, and session replay.
"""

from typing import Optional, Dict, List, Any
from dataclasses import dataclass
import requests


@dataclass
class Message:
    """A message in a conversation session."""
    id: str
    session_id: str
    role: str
    content: str
    sequence_number: int
    created_at: str
    metadata: Optional[Dict[str, Any]] = None

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Message':
        return cls(
            id=data['id'],
            session_id=data['session_id'],
            role=data['role'],
            content=data['content'],
            sequence_number=data['sequence_number'],
            created_at=data['created_at'],
            metadata=data.get('metadata')
        )


@dataclass
class SessionInfo:
    """Session metadata."""
    id: str
    user_id: Optional[str]
    title: Optional[str]
    created_at: str
    updated_at: str
    last_message_at: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'SessionInfo':
        return cls(
            id=data['id'],
            user_id=data.get('user_id'),
            title=data.get('title'),
            created_at=data['created_at'],
            updated_at=data['updated_at'],
            last_message_at=data.get('last_message_at'),
            metadata=data.get('metadata')
        )


class Session:
    """Represents a conversation session with AvocadoDB.

    A session tracks conversation history and provides methods for:
    - Adding user and assistant messages
    - Compiling context for queries
    - Retrieving conversation history
    - Replaying session for debugging

    Example:
        >>> from avocado import AvocadoDB
        >>> db = AvocadoDB()
        >>>
        >>> # Create a new session
        >>> session = db.create_session(user_id="alice", title="My Conversation")
        >>>
        >>> # Compile context and add user message
        >>> result = session.compile("What is Rust?")
        >>> print(result['working_set']['text'])
        >>>
        >>> # Add assistant response
        >>> session.add_message("assistant", "Rust is a systems programming language...")
        >>>
        >>> # Get conversation history
        >>> history = session.get_history()
        >>> print(history)
    """

    def __init__(self, client, session_or_data, session_info: Optional[SessionInfo] = None):
        """Initialize a Session.

        Args:
            client: AvocadoDB client instance
            session_or_data: Session ID (str) or dict with session fields
            session_info: Optional session metadata (ignored if dict is provided)
        """
        self.client = client
        # Allow constructing from dict for tests/convenience
        if isinstance(session_or_data, dict):
            info = SessionInfo.from_dict(session_or_data)
            self.id = info.id
            self.session_id = info.id
            self._info = info
            self.user_id = info.user_id
            self.title = info.title
            self.created_at = info.created_at
            self.updated_at = info.updated_at
        else:
            self.id = str(session_or_data)
            self.session_id = str(session_or_data)
            self._info = session_info
            if session_info is not None:
                self.user_id = session_info.user_id
                self.title = session_info.title
                self.created_at = session_info.created_at
                self.updated_at = session_info.updated_at

    @property
    def info(self) -> SessionInfo:
        """Get session information (lazily loaded)."""
        if self._info is None:
            self._info = self._fetch_info()
        return self._info

    def _fetch_info(self) -> SessionInfo:
        """Fetch session info from server."""
        base_url = getattr(self.client, "url", getattr(self.client, "base_url", None))
        project = getattr(self.client, "project_path", getattr(self.client, "project", "."))
        if base_url:
            response = requests.get(
                f"{base_url}/sessions/{self.session_id}",
                params={"project": project}
            )
            response.raise_for_status()
            data = response.json()
            return SessionInfo.from_dict(data['session'])
        else:
            # CLI mode doesn't support sessions yet
            raise NotImplementedError("Session management not available in CLI mode")

    def add_message(
        self,
        role: str,
        content: str,
        metadata: Optional[Dict[str, Any]] = None
    ) -> Message:
        """Add a message to the session.

        Args:
            role: Message role ('user', 'assistant', 'system', or 'tool')
            content: Message content
            metadata: Optional metadata (tool calls, citations, etc.)

        Returns:
            The created Message object

        Example:
            >>> session.add_message("assistant", "Rust is great!")
        """
        base_url = getattr(self.client, "url", getattr(self.client, "base_url", None))
        project = getattr(self.client, "project_path", getattr(self.client, "project", "."))
        if base_url:
            response = requests.post(
                f"{base_url}/sessions/{self.session_id}/messages",
                json={
                    "role": role,
                    "content": content,
                    "metadata": metadata,
                    "project": project
                }
            )
            response.raise_for_status()
            data = response.json()
            return Message.from_dict(data['message'])
        else:
            raise NotImplementedError("Session management not available in CLI mode")

    def compile(
        self,
        query: str,
        budget: int = 8000,
        **kwargs
    ) -> Dict[str, Any]:
        """Compile context for a query within this session.

        This automatically:
        1. Adds the query as a user message
        2. Compiles the context using the compiler
        3. Associates the working set with the session

        Args:
            query: The user's query
            budget: Token budget for compilation (default: 8000)
            **kwargs: Additional compiler configuration

        Returns:
            Dict with 'message' (Message object) and 'working_set' (WorkingSet)

        Example:
            >>> result = session.compile("How does authentication work?")
            >>> print(result['working_set']['text'])
            >>> print(result['message']['content'])
        """
        base_url = getattr(self.client, "url", getattr(self.client, "base_url", None))
        project = getattr(self.client, "project_path", getattr(self.client, "project", "."))
        if base_url:
            response = requests.post(
                f"{base_url}/sessions/{self.session_id}/compile",
                json={
                    "query": query,
                    "token_budget": budget,
                    "project": project,
                    # TODO: Support additional config from kwargs
                }
            )
            response.raise_for_status()
            data = response.json()

            return data
        else:
            raise NotImplementedError("Session management not available in CLI mode")

    def get_history(self, max_tokens: Optional[int] = None) -> str:
        """Get formatted conversation history.

        Returns the conversation history formatted as:
        ```
        User: <message>

        Assistant: <message>

        User: <message>
        ...
        ```

        Args:
            max_tokens: Optional token limit (keeps most recent messages)

        Returns:
            Formatted conversation history as a string

        Example:
            >>> history = session.get_history(max_tokens=1000)
            >>> print(history)
        """
        base_url = getattr(self.client, "url", getattr(self.client, "base_url", None))
        project = getattr(self.client, "project_path", getattr(self.client, "project", "."))
        if base_url:
            params = {"project": project}
            if max_tokens is not None:
                params["max_tokens"] = max_tokens

            response = requests.get(
                f"{base_url}/sessions/{self.session_id}/history",
                params=params
            )
            response.raise_for_status()
            data = response.json()
            return data['history']
        else:
            raise NotImplementedError("Session management not available in CLI mode")

    def replay(self) -> Dict[str, Any]:
        """Get full session replay for debugging.

        Returns session data with all turns (user message + working set + assistant response).
        Useful for debugging and understanding agent behavior.

        Returns:
            Dict with 'session' and 'turns' (list of conversation turns)

        Example:
            >>> replay = session.replay()
            >>> for turn in replay['turns']:
            ...     print(f"User: {turn['user_message']['content']}")
            ...     if turn['assistant_message']:
            ...         print(f"Assistant: {turn['assistant_message']['content']}")
        """
        base_url = getattr(self.client, "url", getattr(self.client, "base_url", None))
        project = getattr(self.client, "project_path", getattr(self.client, "project", "."))
        if base_url:
            response = requests.get(
                f"{base_url}/sessions/{self.session_id}/replay",
                params={"project": project}
            )
            response.raise_for_status()
            return response.json()
        else:
            raise NotImplementedError("Session management not available in CLI mode")

    def delete(self) -> bool:
        """Delete this session and all associated data.

        Returns:
            True if successful

        Example:
            >>> session.delete()
            True
        """
        base_url = getattr(self.client, "url", getattr(self.client, "base_url", None))
        project = getattr(self.client, "project_path", getattr(self.client, "project", "."))
        if base_url:
            response = requests.delete(
                f"{base_url}/sessions/{self.session_id}",
                params={"project": project}
            )
            response.raise_for_status()
            data = response.json()
            return data['success']
        else:
            raise NotImplementedError("Session management not available in CLI mode")

    def refresh(self) -> None:
        """Refresh this session's metadata from the server."""
        info = self._fetch_info()
        self._info = info
        # Sync common fields for convenience
        self.id = info.id
        self.session_id = info.id
        self.user_id = info.user_id
        self.title = info.title
        self.created_at = info.created_at
        self.updated_at = info.updated_at

    def get_messages(self, limit: Optional[int] = None) -> List[Message]:
        """Get all messages in this session.

        Args:
            limit: Optional limit on number of messages to return

        Returns:
            List of Message objects in chronological order

        Example:
            >>> messages = session.get_messages()
            >>> for msg in messages:
            ...     print(f"{msg.role}: {msg.content}")
        """
        if self.client.mode == "http":
            response = self.client.session.get(
                f"{self.client.url}/sessions/{self.session_id}",
                params={"project": self.client.project_path}
            )
            response.raise_for_status()
            data = response.json()
            messages = [Message.from_dict(m) for m in data['messages']]

            if limit is not None:
                messages = messages[:limit]

            return messages
        else:
            raise NotImplementedError("Session management not available in CLI mode")

    def __repr__(self) -> str:
        user = getattr(self, "user_id", None) or (self._info.user_id if self._info else None) or "N/A"
        return f"Session(id='{self.session_id}', user='{user}')"
